[2/3] Keydata v3 scope changes

sespiros opened the pull request:
Original commit: https://github.com/chrisccoulson/secboot/commit/87d9aa34d1d80d64ffe82f0a008693799d633cd9

I will split the commit above in 3 separate cascading PRs and add tests for it:

1. https://github.com/snapcore/secboot/pull/265
2. This one
3. https://github.com/snapcore/secboot/pull/271

This PR is rebased on top of 1. It adds the changes that split snap model authorization (and boot mode authorization) in a separate package (bootenv) which will now reside outside the cross-platform API.

chrisccoulson commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,488 @@
+	return crypto.Hash(a)
+}
+
+func (a hashAlg) MarshalJSON() ([]byte, error) {
```
There's no test coverage for this or any of the JSON marshalling / unmarshalling functions.

sespiros replied:
Added for hashAlg and ecdsaPublicKey (both MarshalJSON and UnmarshalJSON)

chrisccoulson commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,457 @@
+			role:        "different",
+			validModels: validModels,
+		}), ErrorMatches, "incorrect key supplied")
+}
```
There doesn't appear to be a test of what happens if the wrong key is supplied to `SetAuthorizedSnapModels`.

sespiros replied:
Added.

chrisccoulson commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,457 @@
+		validModes:  validModes,
+		bootMode:    invalidBootMode,
+	}), ErrorMatches, "unauthorized boot mode")
+}
```
Whilst the test coverage here is mostly ok (other than missing the JSON encoding / decoding), it's important to have a test that verifies a specific primary key always derives the same elliptic key - I'll leave it up to you to decide how best to test that, but it will require some hard-coded keys in the test code. There are some examples of this kind of test already (eg, `policyV3SuiteNoTPM.TestDerivePolicyAuthKey*`).

sespiros replied:
Added a test for that too. While adding the tests for marshalling/unmarshalling, I accidentally came across the fact that Go's `ecdsa.GenerateKey` is not deterministic. After adding the test I modified `KeyDataScope.deriveSigner()` to use `ecdsa.GenerateKey()` instead of the internal crypto out of curiosity and my test failed all the times I ran it (I am running 10 iterations). 

Interestingly many of the times it failed even before the test reached my loop, between the 2 derivations that happen in `NewKeyDataScope` itself (one when filling the PublicKey field and one in `authorize()`)

chrisccoulson replied:
Thanks. Yes, go1.20 changed the behaviour of `ecdsa.GenerateKey` to use a non-deterministic method to generate the key, which is a bit of a pain.

The new tests do check that multiple calls to `KeyDataScope.deriveSigner` produce the same key, but there also needs to be a test that verifies it produces the expected key for a fixed input (ie, a test with a hardcoded primary key and expected derived key).

sespiros replied:
I added 2 tests for 2 different fixed keys and another one that tests that different keys are derived from the same primary key but a different role which was also missing.

chrisccoulson commented on keydata_test.go:
```diff
@@ -1427,3 +1562,276 @@ func (s *keyDataSuite) TestReadAndWriteWithLegacySnapModelAuthKey(c *C) {
+	keyDataSuite
+}
+
+func (s *keyDataScopeSuite) SetUpTest(c *C) {
```
As mentioned elsewhere, these kinds of tests should be in the bootenv package.

sespiros replied:
I moved all of them to bootenv. I remember talking about having bootenv potentially as a different package at some point so I copied all the bits I needed from `keydata_test.go` as is. Let me know what you think or if we should create some common package to increase reuse.

chrisccoulson commented on keydata_test.go:
```diff
@@ -448,70 +600,130 @@ func (s *keyDataSuite) checkKeyDataJSONAuthModeNone(c *C, keyData *KeyData, crea
+		pk, err := UnmarshalProtectedKeys(payload)
+		c.Check(err, ErrorMatches, "malformed input")
+		c.Check(pk, IsNil)
+	}
```
Isn't this test only ever called with `s.Version == 2`. In any case, this tests 2 separate functions - it should really be 2 separate tests.

sespiros replied:
This also doesn't exist anymore as it was and it is now split in 2 tests:
`TestKeyPayloadUnmarshalInvalid1` and the new one `TestLegacyKeyPayloadUnmarshalInvalid1`

chrisccoulson reviewed (commented):
I've done another pass and left some comments on test code.

I do think the overall approach for testing here probably needs some tweaks though - I haven't left comments in every place, but I've got some high level notes:
- In general, tests should only be skipped (with `check.C.Skip`) when some pre-requisites are not met (ie, there's something missing in the environment) or if a test is known to fail. It shouldn't be used as a way to exit early from tests to avoid running some combinations in the way that's happening here (ie, `if s.Version == 1`) as it creates entries in the test log that are misleading. The tests should just avoid running invalid combinations entirely.
- I think we should avoid test functions that are used for new and legacy combinations. I would probably update existing test functions to work with the new format (obviously removing the parts and tests that are redundant now), and then add some additional functions just to test some legacy format keys - these can just be a copy of a small subset of existing test code - ie, can we read / unlock / write an existing key, does the legacy snap model authorization work, and can we update the models ok?
- For testing legacy format keys, there is some advantage to just embedding JSON in the tests as opposed to adding code and mocking functions to make it possible to create legacy keys. Running tests on embedded key data provides better guarantees that we don't break them.

Hopefully that makes sense :)

sespiros commented on bootenv/env.go:
```diff
@@ -0,0 +1,39 @@
+	currentBootMode atomic.Value
+)
+
+var SetModel = func(model secboot.SnapModel) bool {
```
The initial branch only allowed `currentModel` and `currentBootMode` to be updated once (using `CompareAndSwap(nil, model)`) so I couldn't change their values between tests. Does that need to change or did I miss something here?

sespiros replied:
I still can't see how I can workaround the fact that atomic.Value can't be set to nil after it is set. My intention here was to mock both the setters and getters to use `keyDataPlatformSuite` variables instead of the real ones. 

I tried changing the setters to use `atomic.Value.Store` then have a new function `ClearModelAndBootMode` in `export_test.go` but `atomic.value.Store`, `atomic.value.Swap` and `atomic.value.CompareAndSwap` operations fail when trying to set a nil value to an atomic variable that has been set.

chrisccoulson replied:
Ah, when I suggested just setting `currentModel` and `currentBootMode` to zero values, I meant not doing it atomically and just doing something like this:
```
currentModel = atomic.Value{}
currentBootMode = atomic.Value{}
```

This obviously isn't safe in production code, but should be fine from test code.

sespiros replied:
Ah right, thanks for this suggestion, it cleaned up the code quite a bit. I implemented it as suggested.

sespiros commented on crypt.go:
```diff
@@ -430,7 +430,10 @@ func ActivateVolumeWithKeyData(volumeName, sourceDevicePath string, authRequesto
+		if keys != nil && keys[0].data.Version == 1 {
+			return errors.New("nil Model")
+		}
+		options.Model = SkipSnapModelCheck
```
Moved the check inside `KeyData.IsSnapModelAuthorized` and added an extra test for this (`cryptLegacySuite.TestActivateVolumeWithKeyDataErrorHandling17`). So basically key data with payload version 1 that fakes its version field to 2 will indeed fail with an unmarshalling error.

sespiros commented on crypt.go:
```diff
@@ -430,7 +430,10 @@ func ActivateVolumeWithKeyData(volumeName, sourceDevicePath string, authRequesto
 		return errors.New("invalid RecoveryKeyTries")
 	}
 	if options.Model == nil {
```
That sounds like a nice cleanup. Separate PR maybe? I will need to check how snapd currently uses this API. So how that would look like from a caller's standpoint, they would need to run `SetModel` first to set the current model right?

chrisccoulson replied:
A separate PR sounds fine. Yes, snap-bootstrap would call `SetModel` and that would delegate to an internal package containing the actual global variables and `loadCurrentModel` etc, which could be consumed by both public packages.

sespiros commented on keydata_file_test.go:
```diff
@@ -131,12 +134,26 @@ func (s *keyDataFileSuite) TestReader(c *C) {
-	authorized, err := keyData.IsSnapModelAuthorized(recoveredAuxKey, models[0])
-	c.Check(err, IsNil)
-	c.Check(authorized, testutil.IsTrue)
+type keyDataFileLegacySuite struct {
```
Removed after the refactor.

sespiros commented on keydata.go:
```diff
@@ -882,7 +676,7 @@ func (d *KeyData) WriteAtomic(w KeyDataWriter) error {
 // ReadKeyData reads the key data from the supplied KeyDataReader, returning a
 // new KeyData object.
-func ReadKeyData(r KeyDataReader) (*KeyData, error) {
+var ReadKeyData = func(r KeyDataReader) (*KeyData, error) {
```
So this should no longer stand as these don't need to be mocked now that I uncoupled the legacy key tests from the new ones but still it is being used in several other places indeed.

For the first point, in most cases that's the issue, that these functions operate on non-exported symbols.
The second point is potentially a good idea, I will keep that in mind, thanks!
I had seen the third approach as well (and had forgotten about it) but I wasn't sure if that was the best way to do it anyway because it felt like a hack :)

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,457 @@
+	_, err := rand.Read(primaryKey)
+	c.Assert(err, IsNil)
+
+	return PrimaryKey(primaryKey)
```
Fixed.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,457 @@
+		ModelAlg:   crypto.SHA256,
+	}
+
+	_, err := NewKeyDataScope(params)
```
Added both Checks.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,457 @@
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+
+	_, err = kds.MakeAdditionalData(1, crypto.SHA256, secboot.AuthModeNone)
```
I committed that as a separate commit as I had to do several changes to do that. Do you have any specific sets of inputs in mind that need to be tested?

sespiros commented on bootenv/export_test.go:
```diff
@@ -0,0 +1,58 @@
+import "github.com/snapcore/secboot"
+
+var (
+	ComputeSnapModelHash = computeSnapModelHash
```
I think I had skipped that because I couldn't think of what could be meaningful to test here. I added one for the valid scenario.

sespiros commented on export_test.go:
```diff
@@ -119,3 +130,112 @@ func MockStderr(w io.Writer) (restore func()) {
+}
+
+// MockMakeDiskUnlockKey uses the new keydata API but creates v1 keydata payloads.
+func MockMakeDiskUnlockKey(primaryKey PrimaryKey) (func(), error) {
```
This doesn't exist anymore, all this functionality got refactored and there is no need to mock these now.

chrisccoulson replied:
Thanks

sespiros commented on export_test.go:
```diff
@@ -119,3 +130,112 @@ func MockStderr(w io.Writer) (restore func()) {
+}
+
+// MockMakeDiskUnlockKey uses the new keydata API but creates v1 keydata payloads.
+func MockMakeDiskUnlockKey(primaryKey PrimaryKey) (func(), error) {
```
This doesn't exist anymore, all this functionality got refactored and there is no need to mock these now.

sespiros commented on keydata_test.go:
```diff
@@ -227,59 +250,138 @@ func toHash(c *C, v interface{}) crypto.Hash {
+		restore, err := MockMakeDiskUnlockKey(primaryKey)
+		c.Assert(err, IsNil)
+		s.AddCleanup(restore)
+	}
```
This doesn't exist anymore, all this functionality got refactored and there is no need to mock these now.

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,488 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
Added doc comments. There was one more, `MakeAdditionalData()` (for platforms without measured boot).

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,488 @@
+	return d.authorize(key, role)
+}
+
+var loadCurrentModel = func() (secboot.SnapModel, error) {
```
Moved.

sespiros commented on keydata_test.go:
```diff
@@ -85,39 +93,54 @@ func (h *mockPlatformKeyDataHandler) unmarshalHandle(data []byte) (*mockPlatform
+	if len(h.scopes) > 0 && !authorized {
+		return nil, err
+	}
+
```
Moved all scope related stuff to bootenv.

chrisccoulson replied:
Thanks

sespiros commented:
@chrisccoulson so regarding your general comment, I reverted the 2 commits that introduced the mixed new/legacy logic for the tests and rewrote them from scratch. I also addressed all of the inline comments.

pedronis commented on bootenv/env.go:
```diff
@@ -0,0 +1,56 @@
+ *
+ */
+
+package bootenv
```
the package itself probably needs a doc comment and maybe a clearer name, I'm not sure why it's called bootenv? it exists to offer key scope support for platforms not doing measured boot?

sespiros replied:
I guess `bootenv` was the initial name because the key scope support is related to the boot environment (boot mode and models). Maybe a different name with the word "scope" in it? `bootscope`/`keyscope`? @chrisccoulson

pedronis replied:
bootscope is not a bad name, adding a package doc comment is still also a good idea, as this is meant to be used by platforms that don't do measurements?

chrisccoulson replied:
I'm fine with bootscope as well.

sespiros replied:
Renamed bootenv to bootscope and added package doc comment.

pedronis commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+	return nil
+}
+
+// MakeAdditionalData constructs the additional data that need to be integrity protected for
+// a key data scope (in AES-GCM for example).
```
this doc comment for this is very opaque, what should be done with the output of this? (the spec doesn't seem very clear on this either)

sespiros replied:
I expanded it a little bit with text from the spec. My understanding is that this serves there as a reference implementation as nothing makes use of it for now. Pinging @chrisccoulson here too :)

pedronis replied:
thanks, now I see that is additional data in the same e.g of https://pkg.go.dev/crypto/cipher@go1.22.0#AEAD ? I wonder though if the name should contain Cipher or AEAD: MakeAEADAdditionalData ?

chrisccoulson replied:
I think `MakeAEADAdditionalData` is probably ok, given that it's expected that another package uses an implementation of `cipher.AEAD` and consumes this.

sespiros replied:
Renamed it as suggested, thanks.

pedronis reviewed (commented):
did a pass, some comments, some questions

sespiros commented on keydata_legacy.go:
```diff
@@ -0,0 +1,321 @@
+	snapModelHMACKDFLabel = []byte("SNAP-MODEL-HMAC")
+)
+
+func unmarshalV1KeyPayload(data []byte) (unlockKey DiskUnlockKey, primaryKey PrimaryKey, err error) {
```
Fixed.

sespiros commented on keydata.go:
```diff
@@ -440,7 +329,7 @@ type keyData struct {
 
 	// AuthorizedSnapModels contains information about the Snap models
 	// that have been authorized to access the data protected by this key.
```
Yes, added expanded the comment a little bit.

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+)
+
+// hashAlg corresponds to a digest algorithm.
+type hashAlg crypto.Hash
```
We do and actually more than twice including the upcoming plainKey PR. See Chris comment below https://github.com/snapcore/secboot/pull/270#discussion_r1500886898

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+	MDAlg  hashAlg `json:"md_alg"`
+}
+
+type AdditionalData struct {
```
Probably not, I made it private again. I believe why I made it public in the first place was because I thought that platform implementations would need to refer to it but I see now that the TPM one brings its own, the upcoming plainkey too. @chrisccoulson am I missing something? Does the spec reflect that?

chrisccoulson replied:
There shouldn't be a need to access this externally - the only thing that is needed externally is the serialized form of it (via `MakeAdditionalData`), and the format of this is a private implementation detail.

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+	default:
+		return 0, errors.New("unsupported hash algorithm")
+	}
+}
```
I moved it in `export_test.go`. If we end up moving `hashAlg` to a separate internal package, it might be good to have that there too.

chrisccoulson replied:
It looks like the only caller of `unmarshalHashAlg` was removed anyway, so you can probably just delete this function now.

sespiros replied:
Deleted, I will keep it locally in case we need it again later though.

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+	MDAlg  hashAlg `json:"md_alg"`
+}
+
+type AdditionalData struct {
```
Samuele also mentioned it and I left a comment there because I had an additional question https://github.com/snapcore/secboot/pull/270#discussion_r1503011853

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+	Role       string
+	KDFAlg     crypto.Hash
+	MDAlg      crypto.Hash
+	ModelAlg   crypto.Hash
```
Added.

sespiros commented on bootenv/keydata.go:
```diff
@@ -0,0 +1,538 @@
+// NewKeyDataScope creates a new scope object from the given parameters.
+//
+// The PrimaryKey and the role parameters are used to derive a role unique
+// signing key which is used to sign a hash (using MDAlg) of an ASN1 marshalled
```
Thanks for pointing that out, it is indeed a bit confusing. I changed several comments to use "DER encoded" where I used "ASN1 marshalled". Is it more accurate?

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+	c.Check(kds, NotNil)
```
Added.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+	c.Check(kds, NotNil)
+
+	err = kds.IsBootEnvironmentAuthorized()
+	c.Check(err, IsNil)
```
Fixed it in 2 places.

sespiros commented on bootenv/export_test.go:
```diff
@@ -0,0 +1,159 @@
+	return pk, nil
+}
+
+func NewPrimaryKey(sz1 int) (secboot.PrimaryKey, error) {
```
Fixed.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+		kds.TestSetVersion(data.keyDataScopeVersion)
+	}
+
+	aadBytes, err := kds.MakeAdditionalData(data.baseVersion, data.keyDigestHashAlg, data.authMode)
```
I generated some fixed ones and replaced what I already had. I tried to think about backwards compatibility, I thought if `UnmarshalAdditionalData` fail for any reason in the future, that would be an indication.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+
+	err = kds.SetAuthorizedSnapModels(primaryKey, data.role, data.validModels...)
```
I originally thought it would be outside the scope of this test but it doesn't hurt to check so I added them.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+	kds, err := NewKeyDataScope(params)
+	c.Assert(err, IsNil)
+
+	err = kds.SetAuthorizedBootModes(primaryKey, data.role, data.validModes...)
```
Added the extra checks here as well.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+		modes:    []string{mode},
+		models:   []SnapModel{modelB},
+	}), ErrorMatches, "cannot perform action because of an unexpected error: unauthorized model")
+}
```
Right, it's like I am testing that the `mockPlatformKeyDataHandler` works :D

When converting the legacy tests, I thought it makes sense to add them here as a reference as iirc I did a single commit where the legacy tests were split in the new ones and these ones and having them there makes it easy for someone looking only at bootenv to make sense of how it is used.

I can remove them if you think they have no use.

chrisccoulson replied:
I think removing them is fine for now. We can add those tests for packages that consume this one instead.

sespiros replied:
Removed.

sespiros commented on keydata_test.go:
```diff
@@ -1048,195 +1064,17 @@ func (s *keyDataSuite) TestChangePassphraseForceIterations(c *C) {
-	c.Check(keyData.ChangePassphrase("passphrase", "12345678", &kdf), Equals, ErrInvalidPassphrase)
-
-	s.checkKeyDataJSONAuthModePassphrase(c, keyData, protected, 0, "12345678", kdfOptions)
-}
```
Oh good catch, thanks. I think I wanted to make it more generic and/or incorporate it in the `testChangePassphrase` group but maybe I missed it. I added it again.

sespiros commented on keydata_test.go:
```diff
@@ -1262,12 +1100,12 @@ func (s *keyDataSuite) TestWriteAtomic1(c *C) {
-	r          KeyDataReader
-	model      SnapModel
-	authorized bool
+	diskUnlockKey DiskUnlockKey
```
Renamed.

sespiros commented on keydata_test.go:
```diff
@@ -1404,6 +1195,250 @@ func (s *keyDataSuite) TestReadKeyData4(c *C) {
+		params:  protected})
+}
+
+func (s *keyDataSuite) TestRecoverLegacyKeyWithPassphrase(c *C) {
```
Removed.

pedronis commented on keydata_test.go:
```diff
@@ -1067,27 +1083,325 @@ func (s *keyDataSuite) TestChangePassphraseWrongPassphrase(c *C) {
-type testSnapModelAuthData struct {
-	alg        crypto.Hash
-	authModels []SnapModel
+func (s *keyDataSuite) TestSnapModelAuthErrorHandling(c *C) {
```
should this test name also contain Legacy?

sespiros replied:
I renamed it and moved it further down under the legacy tests section.

pedronis reviewed (approved):
thanks, replied to some comment threads as well, tentatively approving but would be good to add package comments to bootenv and possibly rename it

chrisccoulson commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1146 @@
+	signer, err := kds.DeriveSigner(primaryKey, role)
+	c.Assert(err, IsNil)
+
+	prevKey, ok := signer.(*ecdsa.PrivateKey)
```
Is this meant to be `privKey`?

sespiros replied:
Yes, fixed.

sespiros commented on bootenv/keydata_test.go:
```diff
@@ -0,0 +1,1008 @@
+
+	err = unmarshalledPk.UnmarshalJSON(pkBytes)
+	c.Check(err, ErrorMatches, "invalid key type")
+}
```
Added `MarshalJSON` and `UnmarshalJSON` for `KeyDataScope` and a test. Should I add fixed ones as well?
